crates/subspace-farmer/src/farming.rs

```diff
@@ -0,0 +1,226 @@
+            })
+            .await?;
+    }
+
+    Ok(())
+}
+
+/// Compare salts in `slot_info` to those known from `salts` and start update plot commitments
+/// accordingly if necessary (in background)
+fn update_commitments(
+    plot: &Plot,
+    commitments: &Commitments,
+    salts: &mut Salts,
+    slot_info: &SlotInfo,
+) {
+    // Check if current salt has changed
+    if salts.current != Some(slot_info.salt) {
+        salts.current.replace(slot_info.salt);
+
+        if salts.next != Some(slot_info.salt) {
+            // If previous `salts.next` is the same as current (expected behavior), need to re-commit
+
+            tokio::spawn({
+                let salt = slot_info.salt;
+                let plot = plot.clone();
+                let commitments = commitments.clone();
+
+                async move {
+                    let started = Instant::now();
+                    info!(
+                        "Salt updated to {}, recommitting in background",
+                        hex::encode(salt)
+                    );
+
+                    if let Err(error) = commitments.create(salt, plot).await {
+                        error!(
+                            "Failed to create commitment for {}: {}",
+                            hex::encode(salt),
+                            error
+                        );
+                    } else {
+                        info!(
+                            "Finished recommitment for {} in {} seconds",
+                            hex::encode(salt),
+                            started.elapsed().as_secs_f32()
+                        );
+                    }
+                }
+            });
+        }
+    }
+
+    if let Some(new_next_salt) = slot_info.next_salt {
+        if salts.next != Some(new_next_salt) {
+            salts.next.replace(new_next_salt);
+
+            tokio::spawn({
+                let plot = plot.clone();
+                let commitments = commitments.clone();
+
+                async move {
+                    let started = Instant::now();
+                    info!(
+                        "Salt will update to {} soon, recommitting in background",
+                        hex::encode(new_next_salt)
+                    );
+                    if let Err(error) = commitments.create(new_next_salt, plot).await {
+                        error!(
+                            "Recommitting salt in background failed for {}: {}",
+                            hex::encode(new_next_salt),
+                            error
+                        );
+                        return;
+                    }
+                    info!(
+                        "Finished recommitment in background for {} in {} seconds",
+                        hex::encode(new_next_salt),
+                        started.elapsed().as_secs_f32()
+                    );
+                }
+            });
+        }
+    }
+}
+
+impl Drop for Farming {
```

Comment:
Please move this next to other impls, it is a bit confusing when impls for the same struct are spaced with so much code.